File: internal/plugins/backends/cursorsdk/protocol/params.go
package protocol

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
)

// ModelSelection is a bridge-private model id (+ optional params).
type ModelSelection struct {
	ID     string          `json:"id"`
	Params json.RawMessage `json:"params,omitempty"`
}

// SandboxOptions is the bridge-private sandbox toggle.
type SandboxOptions struct {
	Enabled bool `json:"enabled"`
}

// AgentCreateLocal holds local-runtime create fields (cwd only at this layer).
type AgentCreateLocal struct {
	Cwd string `json:"cwd"`
}

// ModelsListParams is the models/list request payload (apiKey allowed).
type ModelsListParams struct {
	APIKey string `json:"apiKey"`
}

// AgentCreateParams is the agent/create request payload (apiKey allowed).
// Unknown optional JSON fields are ignored by standard decoding.
type AgentCreateParams struct {
	APIKey             string           `json:"apiKey"`
	Model              ModelSelection   `json:"model"`
	Local              AgentCreateLocal `json:"local"`
	SettingSources     []string         `json:"settingSources"`
	SandboxOptions     *SandboxOptions  `json:"sandboxOptions"`
	AutoReview         bool             `json:"autoReview"`
	EnableAgentRetries bool             `json:"enableAgentRetries"`
	MCPServers         json.RawMessage  `json:"mcpServers,omitempty"`
}

// AgentSendParams is the agent/send request payload (starts a run).
type AgentSendParams struct {
	AgentID string `json:"agentId"`
	Prompt  string `json:"prompt"`
}

// RunCancelParams is the run/cancel request payload.
type RunCancelParams struct {
	RunID string `json:"runId"`
}

// AgentDisposeParams is the agent/dispose request payload.
type AgentDisposeParams struct {
	AgentID string `json:"agentId"`
}

// HealthParams is the bridge/health request payload (empty object).
type HealthParams struct{}

// ShutdownParams is the bridge/shutdown request payload (empty object).
type ShutdownParams struct{}

// DecodeMethodParams validates and decodes typed params for a required method.
// Unknown optional fields are ignored. Structural mandatories are enforced.
// apiKey is allowed only for models/list and agent/create.
func DecodeMethodParams(method string, raw json.RawMessage) (any, error) {
	if len(bytes.TrimSpace(raw)) == 0 {
		raw = json.RawMessage(`{}`)
	}
	if !json.Valid(raw) {
		return nil, protoErr(ErrorInvalidRequest, "params must be a JSON object")
	}
	var obj map[string]json.RawMessage
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	if err := dec.Decode(&obj); err != nil {
		return nil, protoErr(ErrorInvalidRequest, "params must be a JSON object")
	}
	if obj == nil {
		return nil, protoErr(ErrorInvalidRequest, "params must be a JSON object")
	}

	switch method {
	case MethodInitialize:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p InitializeParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.ImplVersion) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing implVersion")
		}
		return p, nil
	case MethodHealth:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p HealthParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		return p, nil
	case MethodModelsList:
		var p ModelsListParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.APIKey) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing apiKey")
		}
		return p, nil
	case MethodAgentCreate:
		for _, key := range []string{"apiKey", "model", "local", "settingSources", "sandboxOptions", "autoReview", "enableAgentRetries"} {
			if _, ok := obj[key]; !ok {
				return nil, protoErr(ErrorInvalidRequest, "missing "+key)
			}
		}
		var p AgentCreateParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.APIKey) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing apiKey")
		}
		if strings.TrimSpace(p.Model.ID) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing model.id")
		}
		if strings.TrimSpace(p.Local.Cwd) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing local.cwd")
		}
		if p.SettingSources == nil {
			return nil, protoErr(ErrorInvalidRequest, "missing settingSources")
		}
		if p.SandboxOptions == nil {
			return nil, protoErr(ErrorInvalidRequest, "missing sandboxOptions")
		}
		if len(p.MCPServers) > 0 && !json.Valid(p.MCPServers) {
			return nil, protoErr(ErrorInvalidRequest, "invalid mcpServers")
		}
		return p, nil
	case MethodAgentSend:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p AgentSendParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.AgentID) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing agentId")
		}
		if strings.TrimSpace(p.Prompt) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing prompt")
		}
		return p, nil
	case MethodRunCancel:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p RunCancelParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.RunID) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing runId")
		}
		return p, nil
	case MethodAgentDispose:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p AgentDisposeParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		if strings.TrimSpace(p.AgentID) == "" {
			return nil, protoErr(ErrorInvalidRequest, "missing agentId")
		}
		return p, nil
	case MethodBridgeShutdown:
		if err := rejectAPIKey(obj); err != nil {
			return nil, err
		}
		var p ShutdownParams
		if err := decodeStrictObject(raw, &p); err != nil {
			return nil, err
		}
		return p, nil
	default:
		return nil, protoErr(ErrorUnknownMethod, method)
	}
}

func rejectAPIKey(obj map[string]json.RawMessage) error {
	if _, ok := obj["apiKey"]; ok {
		return protoErr(ErrorInvalidRequest, "apiKey not allowed for this method")
	}
	return nil
}

func decodeStrictObject(raw json.RawMessage, dest any) error {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	if err := dec.Decode(dest); err != nil {
		return protoErr(ErrorInvalidRequest, err.Error())
	}
	return nil
}

// RedactSecrets removes known secret substrings from diagnostic text.
func RedactSecrets(message string, secrets ...string) string {
	out := message
	for _, secret := range secrets {
		secret = strings.TrimSpace(secret)
		if secret == "" {
			continue
		}
		out = strings.ReplaceAll(out, secret, "[REDACTED]")
	}
	return out
}

// SafeErrorBody builds a protocol error envelope that never echoes secrets.
func SafeErrorBody(code, message string, secrets ...string) ErrorBody {
	return ErrorBody{
		Code:    code,
		Message: RedactSecrets(message, secrets...),
	}
}

// CollectParamSecrets extracts values that must never appear in error text.
func CollectParamSecrets(raw json.RawMessage) []string {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil
	}
	var secrets []string
	if v, ok := obj["apiKey"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil && s != "" {
			secrets = append(secrets, s)
		}
	}
	return secrets
}

// FormatParamError returns a safe invalid_request message for params.
func FormatParamError(method string, raw json.RawMessage, cause error) ErrorBody {
	secrets := CollectParamSecrets(raw)
	msg := fmt.Sprintf("%s params invalid", method)
	if cause != nil {
		msg = msg + ": " + cause.Error()
	}
	return SafeErrorBody(ErrorInvalidRequest, msg, secrets...)
}

File: connectors/cursorsdk/bridge-node/src/params.ts
import { METHODS, type Method, ProtocolError } from "./protocol.js";

export interface InitializeParams {
  implVersion: string;
}

export interface HealthParams {}

export interface ModelsListParams {
  apiKey: string;
}

export interface ModelSelection {
  id: string;
  params?: unknown;
}

export interface SandboxOptions {
  enabled: boolean;
}

export interface AgentCreateLocal {
  cwd: string;
}

export interface AgentCreateParams {
  apiKey: string;
  model: ModelSelection;
  local: AgentCreateLocal;
  settingSources: string[];
  sandboxOptions: SandboxOptions;
  autoReview: boolean;
  enableAgentRetries: boolean;
  mcpServers?: unknown;
}

export interface AgentSendParams {
  agentId: string;
  prompt: string;
}

export interface RunCancelParams {
  runId: string;
}

export interface AgentDisposeParams {
  agentId: string;
}

export interface ShutdownParams {}

function asObject(raw: unknown): Record<string, unknown> {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
    throw new ProtocolError("invalid_request", "params must be a JSON object");
  }
  return raw as Record<string, unknown>;
}

function rejectAPIKey(obj: Record<string, unknown>): void {
  if (Object.prototype.hasOwnProperty.call(obj, "apiKey")) {
    throw new ProtocolError("invalid_request", "apiKey not allowed for this method");
  }
}

function requireString(obj: Record<string, unknown>, key: string): string {
  const v = obj[key];
  if (typeof v !== "string" || v.trim() === "") {
    throw new ProtocolError("invalid_request", `missing ${key}`);
  }
  return v;
}

export function decodeMethodParams(method: Method | string, raw: unknown): unknown {
  const obj = asObject(raw ?? {});
  switch (method) {
    case "bridge/initialize": {
      rejectAPIKey(obj);
      const implVersion = requireString(obj, "implVersion");
      return { implVersion } satisfies InitializeParams;
    }
    case "bridge/health": {
      rejectAPIKey(obj);
      return {} satisfies HealthParams;
    }
    case "models/list": {
      const apiKey = requireString(obj, "apiKey");
      return { apiKey } satisfies ModelsListParams;
    }
    case "agent/create": {
      for (const key of [
        "apiKey",
        "model",
        "local",
        "settingSources",
        "sandboxOptions",
        "autoReview",
        "enableAgentRetries",
      ]) {
        if (!Object.prototype.hasOwnProperty.call(obj, key)) {
          throw new ProtocolError("invalid_request", `missing ${key}`);
        }
      }
      const apiKey = requireString(obj, "apiKey");
      const modelObj = asObject(obj.model);
      const modelId = requireString(modelObj, "id");
      const localObj = asObject(obj.local);
      const cwd = requireString(localObj, "cwd");
      if (!Array.isArray(obj.settingSources)) {
        throw new ProtocolError("invalid_request", "missing settingSources");
      }
      const sandbox = asObject(obj.sandboxOptions);
      if (typeof sandbox.enabled !== "boolean") {
        throw new ProtocolError("invalid_request", "missing sandboxOptions.enabled");
      }
      if (typeof obj.autoReview !== "boolean") {
        throw new ProtocolError("invalid_request", "missing autoReview");
      }
      if (typeof obj.enableAgentRetries !== "boolean") {
        throw new ProtocolError("invalid_request", "missing enableAgentRetries");
      }
      const out: AgentCreateParams = {
        apiKey,
        model: { id: modelId, params: modelObj.params },
        local: { cwd },
        settingSources: obj.settingSources as string[],
        sandboxOptions: { enabled: sandbox.enabled },
        autoReview: obj.autoReview,
        enableAgentRetries: obj.enableAgentRetries,
      };
      if (Object.prototype.hasOwnProperty.call(obj, "mcpServers")) {
        out.mcpServers = obj.mcpServers;
      }
      return out;
    }
    case "agent/send": {
      rejectAPIKey(obj);
      return {
        agentId: requireString(obj, "agentId"),
        prompt: requireString(obj, "prompt"),
      } satisfies AgentSendParams;
    }
    case "run/cancel": {
      rejectAPIKey(obj);
      return { runId: requireString(obj, "runId") } satisfies RunCancelParams;
    }
    case "agent/dispose": {
      rejectAPIKey(obj);
      return { agentId: requireString(obj, "agentId") } satisfies AgentDisposeParams;
    }
    case "bridge/shutdown": {
      rejectAPIKey(obj);
      return {} satisfies ShutdownParams;
    }
    default:
      if (!(METHODS as readonly string[]).includes(method)) {
        throw new ProtocolError("unknown_method", method);
      }
      throw new ProtocolError("invalid_request", `unsupported method ${method}`);
  }
}

export function collectParamSecrets(raw: unknown): string[] {
  try {
    const obj = asObject(raw);
    const key = obj.apiKey;
    return typeof key === "string" && key ? [key] : [];
  } catch {
    return [];
  }
}

export function redactSecrets(message: string, secrets: string[]): string {
  let out = message;
  for (const secret of secrets) {
    if (!secret) continue;
    out = out.split(secret).join("[REDACTED]");
  }
  return out;
}

export function safeErrorBody(
  code: string,
  message: string,
  secrets: string[] = [],
): { code: string; message: string } {
  return { code, message: redactSecrets(message, secrets) };
}

export function formatParamError(
  method: string,
  raw: unknown,
  cause: unknown,
): { code: string; message: string } {
  const secrets = collectParamSecrets(raw);
  const detail = cause instanceof Error ? cause.message : String(cause ?? "");
  return safeErrorBody("invalid_request", `${method} params invalid: ${detail}`, secrets);
}
